File: app/src/main/java/com/shahu/weathery/helper/ValuesConverter.java
package com.shahu.weathery.helper;

import android.annotation.SuppressLint;

import com.shahu.weathery.model.common.MainResponse;

import org.joda.time.DateTime;
import org.joda.time.DateTimeZone;
import org.joda.time.format.DateTimeFormat;
import org.joda.time.format.DateTimeFormatter;

import static com.shahu.weathery.common.Constants.DAY;
import static com.shahu.weathery.common.Constants.NIGHT;

/**
 * Created by Shahu Ronghe on 23, September, 2019
 * in Weathery
 */
public class ValuesConverter {

    private static final String TAG = "ValuesConverter";

    @SuppressLint("DefaultLocale")
    public static String convertTemperatureToCelsius(String temp) {
        //fahrenhiet = (celsius * 9.0/5.0) + 32.0;
        double celsius, kelvin;
        kelvin = Double.parseDouble(temp);
        celsius = kelvin - 273.0;
        return String.format("%.0f", celsius);
    }

    public static DateTime convertUnixTime(long epoch, int secShift) {
        DateTime dateTime = new DateTime((long) epoch * 1000);
        dateTime = dateTime.withZone(DateTimeZone.UTC);
        dateTime = dateTime.plusSeconds(secShift);
        return dateTime;
    }

    public static String getDayNight(MainResponse mainResponse) {
        int secShift = mainResponse.getTimezone();
        int sunrise = convertUnixTime(mainResponse.getSys().getSunrise(), secShift).getHourOfDay();
        int sunset = convertUnixTime(mainResponse.getSys().getSunset(), secShift).getHourOfDay();
        int current = convertUnixTime(mainResponse.getDt(), secShift).getHourOfDay();
        if (current > sunrise && current < sunset)
            return DAY;
        return NIGHT;
    }

    public static String getCountryImage(String countryCode) {
        try {
            int flagOffset = 0x1F1E6;
            int asciiOffset = 0x41;
            int firstChar = Character.codePointAt(countryCode, 0) - asciiOffset + flagOffset;
            int secondChar = Character.codePointAt(countryCode, 1) - asciiOffset + flagOffset;

            return new String(Character.toChars(firstChar))
                    + new String(Character.toChars(secondChar));
        } catch (Exception e) {
            return null;
        }
    }

    public static String getTimeForCity(long time, int secondsShift) {
        String newTime = null;
        DateTime dateTime = convertUnixTime(time,secondsShift);
        DateTimeFormatter dateTimeFormatter = DateTimeFormat.forPattern("E, HH:mm");
        newTime =  dateTimeFormatter.print(dateTime);
        return newTime;
    }
}

File: app/src/main/java/com/shahu/weathery/helper/ValuesConverter.kt
package com.shahu.weathery.helper

import android.annotation.SuppressLint
import com.shahu.weathery.common.Constants
import org.joda.time.DateTime
import org.joda.time.DateTimeZone
import org.joda.time.format.DateTimeFormat

/**
 * Created by Shahu Ronghe on 23, September, 2019
 * in Weathery
 */
object ValuesConverter {
    private const val TAG = "ValuesConverter"

    @JvmStatic
    @SuppressLint("DefaultLocale")
    fun convertTemperatureToCelsius(temp: String): String {
        //fahrenhiet = (celsius * 9.0/5.0) + 32.0;
        val celsius: Double
        val kelvin: Double = temp.toDouble()
        celsius = kelvin - 273.0
        return String.format("%.0f", celsius)
    }

    private fun convertUnixTime(epoch: Long, secShift: Int): DateTime {
        var dateTime = DateTime(epoch * 1000)
        dateTime = dateTime.withZone(DateTimeZone.UTC)
        dateTime = dateTime.plusSeconds(secShift)
        return dateTime
    }

    fun getDayNight(timezone: Int, sunriseTime: Int, sunsetTime: Int, dtIncoming: Int): String {
        val sunrise = sunriseTime.toLong().let { convertUnixTime(it, timezone).hourOfDay }
        val sunset = sunsetTime.toLong().let { convertUnixTime(it, timezone).hourOfDay }
        val current = convertUnixTime(dtIncoming.toLong(), timezone).hourOfDay
        return if (current in (sunrise.plus(1)) until sunset) Constants.DAY else Constants.NIGHT
    }

    @JvmStatic
    fun getCountryImage(countryCode: String?): String? {
        return try {
            val flagOffset = 0x1F1E6
            val asciiOffset = 0x41
            val firstChar = Character.codePointAt(countryCode, 0) - asciiOffset + flagOffset
            val secondChar = Character.codePointAt(countryCode, 1) - asciiOffset + flagOffset
            (String(Character.toChars(firstChar))
                    + String(Character.toChars(secondChar)))
        } catch (e: Exception) {
            null
        }
    }

    @JvmStatic
    fun getTimeForCity(time: Long, secondsShift: Int): String {
        val dateTime = convertUnixTime(time, secondsShift)
        val dateTimeFormatter = DateTimeFormat.forPattern("E, HH:mm")
        return dateTimeFormatter.print(dateTime)
    }

    @JvmStatic
    fun getDateForCity(currentTime: Long, timezone: Int): String {
        val dateTime = convertUnixTime(currentTime, timezone)
        val dateTimeFormatter = DateTimeFormat.forPattern("EEEE, dd MMM")
        return dateTimeFormatter.print(dateTime)
    }

    @JvmStatic
    fun getTimeOnlyForCity(currentTime: Long, timezone: Int): String {
        val dateTime = convertUnixTime(currentTime, timezone)
        val dateTimeFormatter = DateTimeFormat.forPattern("HH:mm")
        return dateTimeFormatter.print(dateTime)
    }

    fun getHourOfTheDayByEpoch(time: Long, timezone: Int): String? {
        val dateTime = convertUnixTime(time, timezone)
        val dateTimeFormatter = DateTimeFormat.forPattern("HH")
        return dateTimeFormatter.print(dateTime)
    }

    fun getDayOfTheWeek(time: Int, timezone: Int): String? {
        val dateTime = convertUnixTime(time.toLong(), timezone)
        val dateTimeFormatter = DateTimeFormat.forPattern("EEEE")
        return dateTimeFormatter.print(dateTime)
    }
}